### ISSUE TITLE
Safe transaction execution: Incosistency/unclear behavior between UI and SDK

### ISSUE THREAD

**Issue by perplexedpigmy** [safe-global/safe-core-sdk#319]
Working on a safe requiring 3 confirmations.
When a transaction is proposed via the SDK it has 1 confirmation
```
  confirmations: [
    {
      owner: '0x5b0eC2a748E4ED54329b11e192fA0594849531e5',
      submissionDate: '2023-01-03T14:54:18.256432Z',
      transactionHash: null,
      signature: '0xfbbcdc14cb37334392a46c5233b241c4eabda9979d48fcea0c2f26fe429bb10674a05296b3c7aab84a30a131cb799801906b8381254258f93a4f549890b8408d20',
      signatureType: 'ETH_SIGN'
    },
  ],
```

After adding 2 additional confirmations  (via the SDK)
```
[
...
    {
      owner: '0x97bC34017B9d6F356D58D670d7E6495190607Ca2',
      submissionDate: '2023-01-03T15:30:00Z',
      transactionHash: '0x0db2443034efe574d6022f64c9f6b3f258fc741cda6c36d358aad47cef5f1fa1',
      signature: '0x00000000000000000000000097bc34017b9d6f356d58d670d7e6495190607ca2000000000000000000000000000000000000000000000000000000000000000001',
      signatureType: 'APPROVED_HASH'
    },
    {
      owner: '0x3067338Ce95737e7a1a328E62e836613150e0c09',
      submissionDate: '2023-01-03T15:30:24Z',
      transactionHash: '0x64b67e85bed2c41561963c3012fc2b6f7c7b37dd634eac63b73ce6b4287d851f',
      signature: '0x0000000000000000000000003067338ce95737e7a1a328e62e836613150e0c09000000000000000000000000000000000000000000000000000000000000000001',
      signatureType: 'APPROVED_HASH'
    }
]
``` 
The safe transaction has 2 + 1 = 3 confirmations, and the UI is happy to enable the execute button.
![The UI allows enables execution](https://user-images.githubusercontent.com/9532454/210412130-c0edd194-0a54-4482-9f83-0de3751e4c3d.png)
because it counts the number of confirmations from the Safe transaction information

While the SDK reads the number of [confirmations from Ethereum ](https://github.com/safe-global/safe-core-sdk/blob/de8d2b9e4c279142f3b878f938877f846886bea3/packages/safe-ethers-lib/src/contracts/GnosisSafe/GnosisSafeContractEthers.ts#L59) and gets only the number of confirmations committed to Ethereum. Which in this case are only the last two.

Causing a call to the  SDK 's `Safe.executeTransaction`  
to yield an `ExecuteError: There is 1 signature missing`

I think the UI's logic makes more sense, but in any case it's better if there is a consistent behavior so we know how to configure the safe.

**Comment by katspaugh** [comment]
I agree, and I think it's important to fix this bug.

**Comment by perplexedpigmy** [comment]
The plot thickens, in an effort to find a possible some way to have SDK and UI consistency, I tried the `delegate proposer` way. 
The threshold was decreased to two and the proposal was created via a delegate yielding a safe transaction with an empty confirmation array

Signing it using the SDK gives.
```
  confirmations: [
    {
      owner: '0x97bC34017B9d6F356D58D670d7E6495190607Ca2',
      submissionDate: '2023-01-04T19:49:36Z',
      transactionHash: '0x423946e939b154a00699e653834b795d06de95a30166454f1e9d5844355e5dbc',
      signature: '0x00000000000000000000000097bc34017b9d6f356d58d670d7e6495190607ca2000000000000000000000000000000000000000000000000000000000000000001',
      signatureType: 'APPROVED_HASH'
    },
    {
      owner: '0x3067338Ce95737e7a1a328E62e836613150e0c09',
      submissionDate: '2023-01-04T19:50:00Z',
      transactionHash: '0x07a237babc73cc933c0431d99eaf6b515fbefd86e13edaaefe65c0b92ddd9c8f',
      signature: '0x0000000000000000000000003067338ce95737e7a1a328e62e836613150e0c09000000000000000000000000000000000000000000000000000000000000000001',
      signatureType: 'APPROVED_HASH'
    },
]
```

While the expectation is that the execution will work the result were surprisingly  `ExecuteError: There is 1 signature missing`
A closer look, shows that [the same code mentioned above](https://github.com/safe-global/safe-core-sdk/blob/de8d2b9e4c279142f3b878f938877f846886bea3/packages/safe-ethers-lib/src/contracts/GnosisSafe/GnosisSafeContractEthers.ts#L60) doesn't return any Owners approving transactions.

Worse, now the execute button is also disabled on the UI
And in an effort to have the button enabled, another signature was added , to no avail, 3 signature out of required 2 and the execution is still disabled.
![image](https://user-images.githubusercontent.com/9532454/210647250-53a13dac-20b2-4fe0-88db-69d5d9034f02.png)

I stumbled into a consistent case where both UI and SDK fail to provide a way to execute the transaction, albeit I was looking for a more positive consistent result

**Comment by germartinez** [comment]
> Causing a call to the SDK 's `Safe.executeTransaction`
> to yield an `ExecuteError: There is 1 signature missing`

Hi @perplexedpigmy,
Could you please write down all the steps to get this error? Specifying what signers do you connect to the SDK, and the method you use to sign. Also indicating who is the account executing the transaction with the SDK. That way I can try to replicate it.
Thank you!

**Comment by germartinez** [comment]
> While the expectation is that the execution should work, the result were surprisingly `ExecuteError: There is 1 signature missing`

Same here :)

**Comment by perplexedpigmy** [comment]
> > Causing a call to the SDK 's `Safe.executeTransaction`
> > to yield an `ExecuteError: There is 1 signature missing`
> 
> Hi @perplexedpigmy, Could you please write down...

Thanks @germartinez 
False alarm, Was too trigger happy thinking I am providing more information for the same problem.
it was a local issue  that evaporated after a restart, with UI behavior I can think of some plausible browser explanations, with the SDK, I have nothing, and so far it's not reproducible.

For the original problem, there is a discrepancy but it can be worked around (Or it might have been the intention of the author, albeit I didn't find any such documentation)
the [code](https://github.com/safe-global/safe-core-sdk/blob/3327b7b3e51e5f1abe9b1a7c44f8695ad32aa5a5/packages/safe-core-sdk/src/Safe.ts#L851)  is reading all confirmations from the chain, and then opts to add the `Safe.executeTransaction` signer as another confirmations, 

If the signer is in the list of signers retrieved from the contract the execution fails, otherwise it's added to the number of confirmations and the execution is successful, 

In the initial comment I posted it's enough to execute the transaction using `0x5b0eC2a748E4ED54329b11e192fA0594849531e5` to have a successful execution.

Bottom line a delegate or the workaround are sufficient to have consistent behavior, but it will be nice to have the UI/SDK work exactly the same way.

**Comment by germartinez** [comment]
The `safe-core-sdk` package does not use the Safe Transaction Service, where the transactions/signatures are published to be shared among the signers and the one consumed by the UI, as it is an SDK that has to work with anything but the contracts. For that reason, signatures that are sent to the service are not detected by the SDK, only the ones added locally or confirmed on-chain are.

However, the `safe-service-client` package (inside the Safe Core SDK monorepo as well) consumes the Safe Transaction Service API. If the transaction is retrieved from this service and executed with the `safe-core-sdk`, it should contain all the signatures.

```js
import Safe from '@safe-global/safe-core-sdk'
import SafeServiceClient from '@safe-global/safe-service-client'

// Here it goes the initialization of the SDKs...

// Get a transaction from the service
const safeTransaction = await safeService.getTransaction('<SAFE_TX_HASH>')

// Execute the transaction with the SDK (it has signatures from service + on-chain + executor (if signer))
const executeTxResponse = await safeSdk.executeTransaction(safeTransaction)
```

Here is the guide on how to use both SDKs combined: https://github.com/safe-global/safe-core-sdk/blob/main/guides/integrating-the-safe-core-sdk.md#8-execute-the-transaction.

Does your code look like that one?

**Comment by perplexedpigmy** [comment]
Excellent explanation, 

There is my problem, I was executing using SafeHashId not the transaction itself.
Something to the effect of 

```javascript
for ( const multiSigTransactionResponse of (await SafeServiceClient.getPendingTransaction(SAFE_ADDR).results) { 
   const executeTxResponse = await safeSdk.executeTransaction(multiSigTransactionResponse.safeTxHash, EXECUTER_KEY)
}
```

**Comment by germartinez** [comment]
I see. In that case I think this issue can be closed. Let me know if I'm missing anything.

**Comment by germartinez** [comment]
Hi @omeguhh, are you using the `safe-core-sdk` `v3.2.2`?
It should accept both types: https://github.com/safe-global/safe-core-sdk/blob/main/packages/safe-core-sdk/src/Safe.ts#L841

**Comment by germartinez** [comment]
`getOwners` can't be called with `executeTransaction` method. It should be called this way:
```js
const ownerAddresses = await safeSdk.getOwners()
```
In the Safe contracts, `getOwners` is a public function that just reads the storage: https://github.com/safe-global/safe-contracts/blob/da66b45ec87d2fb6da7dfd837b29eacdb9a604c5/contracts/base/OwnerManager.sol#L136. The are no transactions involved, it is just a call.

**Comment by germartinez** [comment]
Can you copy-paste the Safe transaction you are trying to execute? Or share your Safe address if it's not private?

**Comment by germartinez** [comment]
I tried calling the method `isValidTransaction` with your transaction and I got `true`.
I suggest you check this script: https://github.com/safe-global/safe-core-sdk/blob/main/playground/execute-transaction.ts
Update the `config` object with the right data for `SIGNER_ADDRESS_PRIVATE_KEY` and `INFURA_KEY`.
```js
const config: Config = {
  RPC_URL: 'https://goerli.infura.io/v3/<INFURA_KEY>',
  SIGNER_ADDRESS_PRIVATE_KEY: '<PRIVATE_KEY>',
  SAFE_ADDRESS: '0xcA1E06fE5416F192Bd9c1194c472511edaEA58F4',
  TX_SERVICE_URL: 'https://safe-transaction-goerli.safe.global/',
  SAFE_TX_HASH: '0x6f2115a8ac03dbbbb45a664234a941de3d21d9ab21d73791449f387dcdde379c'
}
```
Then execute the script with the following command:
```
yarn play execute-transaction
```

**Comment by germartinez** [comment]
I just tested with Alchemy RPC and `isValidTransaction` returns `true` as well.

```
https://eth-goerli.g.alchemy.com/v2/<ALCHEMY_KEY>
```
